core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java

```diff
@@ -0,0 +1,190 @@
+            Address primary = readCH.locatePrimaryOwnerForSegment(segmentId);
+            List<Address> readOwners = readOwnersMap.computeIfAbsent(primary, p ->
+                  Immutables.immutableListCopy(readCH.locateOwnersForSegment(segmentCopy)));
+            List<Address> writeOwners = writeOwnersMap.computeIfAbsent(primary, p ->
+                  Immutables.immutableListCopy(writeCH.locateOwnersForSegment(segmentCopy)));
+            List<Address> writeBackups = writeOwners.subList(1, writeOwners.size());
+            this.distributionInfos[segmentId] =
+                  new DistributionInfo(segmentId, primary, readOwners, writeOwners, writeBackups, localAddress);
+         }
+      } else { // Invalidation/Local
+         assert cacheMode.isInvalidation() || cacheMode == CacheMode.LOCAL;
+         // Reads and writes are local, only the invalidation is replicated
+         Address primary = localAddress;
+         List<Address> owners = Collections.singletonList(localAddress);
+         List<Address> writeBackups = Collections.emptyList();
+         this.distributionInfos = new DistributionInfo[]{
+               new DistributionInfo(0, primary, owners, owners, writeBackups, localAddress)
+         };
+      }
+   }
+
+   /**
+    * @return {@code true} iff key {@code key} can be read without going remote.
+    */
+   public boolean isReadOwner(Object key) {
+      // In replicated and invalidation mode only the primary owner changes from one segment to the next
+      int segmentId = isDistributed ? keyPartitioner.getSegment(key) : 0;
+      return distributionInfos[segmentId].isReadOwner();
+   }
+
+
+   /**
+    * @return {@code true} iff writing a value for key {@code key} will update it on the local node.
+    */
+   public boolean isWriteOwner(Object key) {
+      // In replicated and invalidation mode only the primary owner changes from one segment to the next
+      int segmentId = isDistributed ? keyPartitioner.getSegment(key) : 0;
+      return distributionInfos[segmentId].isWriteOwner();
+   }
+
+   /**
+    * @return The consistent hash segment of key {@code key}
+    */
+   public int getSegment(Object key) {
+      return keyPartitioner.getSegment(key);
+   }
+
+   /**
+    * @return Information about the ownership of segment {@code segment}, including the primary owner.
+    */
+   public DistributionInfo getDistributionForSegment(int segmentId) {
+      return distributionInfos[segmentId];
+   }
+
+   /**
+    * @return Information about the ownership of key {@code key}, including the primary owner.
+    */
+   public DistributionInfo getDistribution(Object key) {
+      int segmentId = (isDistributed || isReplicated) ? keyPartitioner.getSegment(key) : 0;
+      return distributionInfos[segmentId];
+   }
+
+   /**
+    * @return An unordered collection with the write owners of {@code key}, or {@code null} if the cache is replicated.
+    */
+   public Collection<Address> getWriteTargets(Object key) {
```

Comment:
I wanted to keep it here so that it's available for `ClusteringDependentLogic.getOwners()`. But in the meantime I decided that `ClusteringDependentLogic` was only accessible via injection or `AdvancedCache.getComponentRegistry()`, so I can remove methods without breaking backwards compatibility.